File: jmazes/src/main/java/com/landsmann/mazes/demos/DeadendCount.java
package com.landsmann.mazes.demos;

import com.landsmann.mazes.algorithms.*;
import com.landsmann.mazes.core.Grid;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.*;

public class DeadendCount {

    public static void main(String[] args) {
        List<Class<?>> algorithms = Arrays.asList(BinaryTree.class, Sidewinder.class,
                AldousBroder.class, Wilsons.class, HuntAndKill.class, RecursiveBackTracker.class);

        int tries = 100;
        int size = 20;

        Map<Class, Integer> averages = new HashMap<>();
        algorithms.forEach(algorithm -> {
            System.out.println("running " + algorithm.getSimpleName());

            List<Integer> deadendCounts = new ArrayList<Integer>();

            for (int i = 0; i < tries; i++) {
                Grid grid = new Grid(size, size);
                try {
                    Method on = algorithm.getMethod("on", Grid.class);
                    on.invoke(null, grid);
                    deadendCounts.add(grid.deadends().size());
                } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
                    e.printStackTrace();
                }
            }

            int totalDeadends = deadendCounts.stream()
                    .mapToInt(i -> i)
                    .sum();
            averages.put(algorithm, totalDeadends / deadendCounts.size());
        });

        int totalCells = size * size;
        System.out.println();
        System.out.println("Average dead-end per " + size + "x" + size + " maze (" + totalCells + ") cells:");
        System.out.println();

        algorithms.sort((a1, a2) -> averages.get(a2).compareTo(averages.get(a1)));

        algorithms.forEach(algorithm -> {
            int percentage = (int) (averages.get(algorithm) * 100f / (size * size));
            System.out.println(
                    String.format("%1$s : %2$d/%3$d (%4$d%%)",
                            algorithm.getSimpleName(),
                            averages.get(algorithm),
                            totalCells,
                            percentage));
        });
    }
}

File: ruby_mazes/deadend_count.rb
require 'grid'
require 'binary_tree'
require 'sidewinder'
require 'aldous_broder'
require 'wilsons'
require 'hunt_and_kill'
require 'recursive_back_tracker'

algorithms = [BinaryTree, Sidewinder, AldousBroder, Wilsons, HuntAndKill, RecursiveBackTracker]

tries = 100
size = 20

averages = {}
algorithms.each do |algorithm|
  puts "running #{algorithm}"

  deadend_counts = []
  tries.times do
    grid = Grid.new(size, size)
    algorithm.on(grid)
    deadend_counts << grid.deadends.count
  end

  total_deadends = deadend_counts.inject(0) { |s,a| s + a }
  averages[algorithm] = total_deadends / deadend_counts.length
end

total_cells = size * size
puts
puts "Average dead-end per #{size}x#{size} maze (#{total_cells} cells):"
puts

sorted_algorithms = algorithms.sort_by { |algorithm| -averages[algorithm]}

sorted_algorithms.each do |algorithm|
  percentage = averages[algorithm] * 100.0 / (size * size)
  puts "%14s : %3d/%d (%d%%)" %
      [algorithm, averages[algorithm], total_cells, percentage]
end

